src/ensembl/src/content/app/browser/browser-region-field/BrowserRegionField.tsx

```diff
@@ -0,0 +1,226 @@
+};
+
+export const BrowserRegionField = (props: BrowserRegionFieldProps) => {
+  const { activeGenomeId } = props;
+  const [regionFieldInput, setRegionFieldInput] = useState('');
+  const [errorMessages, setErrorMessages] = useState<string | null>(null);
+  const [isFormSubmitted, setIsFormSubmitted] = useState(false);
+
+  const activateForm = () => {
+    if (!props.regionFieldActive && !props.isDrawerOpened) {
+      props.toggleRegionFieldActive(true);
+    }
+  };
+
+  const changeRegionFieldInput = (value: string) => setRegionFieldInput(value);
+
+  const getRegionInputWithRegion = (input: string) => {
+    const [region, ,] = props.chrLocation;
+
+    return input.includes(':') ? input : `${region}:${input}`;
+  };
+
+  const closeForm = () => {
+    setRegionFieldInput('');
+    setIsFormSubmitted(false);
+    setErrorMessages(null);
+
+    props.toggleRegionFieldActive(false);
+    props.resetRegionValidation();
+  };
+
+  const handleSubmit = (event: FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    if (activeGenomeId && regionFieldInput && props.chrLocation) {
+      setIsFormSubmitted(true);
+      props.validateRegion(getRegionInputWithRegion(regionFieldInput));
+    }
+  };
+
+  const changeLocation = (inputValue: string) => {
+    if (!isFormSubmitted) {
+      return;
+    }
+
+    props.changeBrowserLocation(
+      props.activeGenomeId as string,
+      getChrLocationFromStr(getRegionInputWithRegion(inputValue))
+    );
+  };
+
+  const changeFocusObject = (inputValue: string) => {
+    const activeEnsObjectId = `${
+      props.activeGenomeId
+    }:region:${getRegionInputWithRegion(inputValue)}`;
+
+    const params = {
+      genomeId: props.activeGenomeId,
+      focus: activeEnsObjectId,
+      location: getRegionInputWithRegion(inputValue)
+    };
+
+    props.replace(urlFor.browser(params));
+  };
+
+  const updateBrowser = () => {
+    if (!isFormSubmitted) {
+      return;
+    }
+
+    const [region, ,] = props.chrLocation;
+    const regionInput = regionFieldInput.includes(':')
+      ? regionFieldInput.split(':')[0]
+      : region;
+
+    if (regionInput === region) {
+      changeLocation(regionFieldInput);
+    } else {
+      changeFocusObject(regionFieldInput);
+    }
+  };
+
+  useEffect(() => () => props.resetRegionValidation(), []);
+
+  useEffect(() => {
+    if (props.regionFieldActive) {
+      const { regionValidationInfo, regionValidationLoadingStatus } = props;
+
+      if (regionValidationLoadingStatus !== LoadingState.LOADING) {
+        const errorMessages = getRegionFieldErrorMessages(regionValidationInfo);
+
+        if (errorMessages) {
+          setErrorMessages(errorMessages);
+        } else {
+          setRegionFieldInput('');
+          setErrorMessages(null);
+          setIsFormSubmitted(false);
+          props.toggleRegionFieldActive(false);
+
+          updateBrowser();
+        }
+      }
+    }
+  }, [props.regionValidationInfo]);
```

Comment:
Same as for region editor — update browser from within react action after successful response?

Reply:
I have changed the code. Is it better now?

Reply:
Yes, it’s gotten much simpler and easier to reason about :smiley: